packages/client-utils/src/vstorage-client.js

```diff
@@ -0,0 +1,489 @@
+
+      return BigInt(result.sync_info.latest_block_height);
+    }
+
+    throw Error(
+      `Couldn't get latest height from any of the rpcs in ${quote(config.rpcAddrs)}`,
+    );
+  };
+
+  /**
+   * @overload
+   * @param {string} path
+   * @param {Partial<{ height: Height; kind: PATHS['CHILDREN'] }>} opts
+   * @returns {Promise<QueryChildrenResponse>}
+   *
+   * @overload
+   * @param {string} path
+   * @param {Partial<{ height: Height; kind: PATHS['DATA'] }>} opts
+   * @returns {Promise<QueryDataResponse>}
+   */
+  /**
+   * @param {string} path
+   * @param {Partial<{ height: Height; kind: PATHS['CHILDREN'] | PATHS['DATA'] }>} [opts]
+   * @returns {Promise<QueryChildrenResponse | QueryDataResponse>}
+   */
+  const queryAbci = async (
+    path,
+    { height = 0n, kind = PATHS.CHILDREN } = {},
+  ) => {
+    await null;
+
+    const codec = codecs[kind];
+    /** @type {Response|undefined} */
+    let response;
+    const route = createVstorageAbciRoute(height, kind, path);
+    const rpcAddress = getRpcAddress();
+
+    const url = rpcAddress + route;
+
+    try {
+      response = await fetch(url, { keepalive: true });
+    } catch (err) {
+      console.error(err);
+      throw createError(err.message, height, kind, path, rpcAddress);
+    }
+
+    if (!response.ok)
+      throw createError(await response.text(), height, kind, path, rpcAddress);
+
+    const { result } = /** @type {AbciResponse} */ (await response.json());
+
+    if (result.response?.code !== 0)
+      throw createError(
+        `Error code '${result.response.code}', Error message: '${result.response.log}'`,
+        height,
+        kind,
+        path,
+        rpcAddress,
+      );
+
+    return codec.response.decode(decodeBase64(result.response.value));
+  };
+
+  return { getLatestHeight, queryAbci };
+};
+
+/**
+ * @template {any} T
+ * @param {Array<T>} values
+ */
+const makeRoundRobin = values => {
+  let currentIndex = 0;
+
+  return () => {
+    const element = values[currentIndex];
+    currentIndex = (currentIndex + 1) % values.length;
+    return element;
+  };
+};
+
+/**
+ * @template {any} T
+ * @param {{queryClient: ReturnType<typeof makeQueryClient>}} powers
+ * @param {string} path
+ * @param {Partial<{ compat: boolean }>} [options]
+ * @returns {StreamTopic<T>}
+ */
+const makeStreamTopic = ({ queryClient }, path, options) => {
+  /**
+   * @param {Height} blockHeight
+   * @param {StreamCell['values']} values
+   */
+  function* allValuesFromCell(blockHeight, values) {
+    for (const value of values.reverse())
+      yield /** @type {Update<T>} */ ({
+        blockHeight,
+        value: JSON.parse(value),
+      });
+  }
+
+  /**
+   * @param {boolean} compatMode
+   * @param {ReturnType<typeof parseValue>} value
+   */
+  const assertValueStructure = (compatMode, value) => {
+    if (!(isStreamCell(value) || compatMode))
+      throw Error(`Expected a stream cell, got ${quote(value)}`);
```

Comment:
There’s an opportunity to take advantage of short-circuiting here to avoid extra work in non-compat-mode. First, _De Morgan_ the expression then reverse the order of operands. The result is logically equivalent but faster in the non-compat case.
```suggestion
        if (!compatMode && !isStreamCell(value))
      throw Error(`Expected a stream cell, got ${quote(value)}`);
```